### PR TITLE
Miscellaneous improvements to splitting

### PR DESCRIPTION
**Pull request by MaLoefUDS**
### Prerequisites

- [x] I adhere to the coding conventions (described [here](../CONTRIBUTING.md)) in my code.
- [x] I have updated the copyright headers of the files I have modified.
- [x] I have written appropriate commit messages, i.e., I have recorded the goal, the need, the needed changes, and the location of my code modifications for each commit. This includes also, e.g., referencing to relevant issues.
- [x] I have put signed-off tags in *all* commits.
- [x] I have updated the changelog file [NEWS.md](../NEWS.md) appropriately.
- [x] I have checked whether I need to adjust the showcase file `showcase.R` with respect to my changes.
- [x] The pull request is opened against the branch `dev`.

### Description

This PR is targeted at enhancing several splitting related functionalities listed in #187. It simplifies network splitting, fixes a bug in activity-based network-, and data-splitting, and extends the coronet test suite.

### Changelog

Added:

* Deduplication of issue-based artifact-network-edges, that are a result of code-faces extraction (and this [PR](https://github.com/se-sic/codeface-extraction/pull/51))
* Introduced a minor wrapper function
* Validate the presence and correctness of the 'bins' attribute on split data and networks

Changed:

* Setting the "bins" attribute in time-based network-splitting in fewer places
* Miscellaneous code simplifications
* Apply the same algorithm in the sliding-window approach of network-splitting, as in data-splitting

Fixed:

* Fixed a bug in the logical, that decides wether the last sliding-window or regular-window is redundant and can be removed, in both data-, and network-splitting
* Minor fix in a test
* Ensures the presence of the 'bins' attribute on network splits produced by any splitting function

### CODE REVIEW

**Comment by MaLoefUDS** [comment]
I have also worked on adding checks for the `bins` in network-splitting tests. Adding a check is not hard, its easily doable but I have one question regarding it. 

Splitting time-based does assume the bins attribute to be of type `POSIXct`, while activity-based splitting does assume it to be of type `character`. Is this okay, or should I level it, if yes, on what? (Also, there is exactly one test that already tests for the bins parameter for time-based network-splitting, it assumes `POSIXct` of course).

Maybe as a decision helper: In data-splitting the `bins` attribute is always assumed to be of type `characer`, in both time-based and activity-based splitting.

**Comment by MaLoefUDS** [comment]
Also, I have reviewed, how many tests fail when I add a single duplicate JIRA edge. We get 66 failed tests from a multitude of files, similar to the other changes on the testing data I did previously, just in a smaller scale.

**Comment by MaLoefUDS** [comment]
I manually installed igraph version 2.0.1.1 and ran the tests locally. For me there are no failing tests anymore 🤔. The CI also fetches igraph version 2.0.1.1, so I don't really know what is going on here.

File: util-split.R [hunk 1]
```diff
@@ -739,33 +731,41 @@ split.network.activity.based = function(network, number.edges = 5000, number.win
...
-        ## remove the last regular range as it is not complete and we don't loose data when removing it
-        edges.last.regular = igraph::E(networks[[length(networks)]])
-        edges.last.sliding = igraph::E(networks[[length(networks) - 1]])
-        if (bins.date[length(bins.date)] == bins.date.middle[length(bins.date.middle)]
-            && all(edges.last.regular %in% edges.last.sliding)
-            && table(edges.last.regular$date) %in% table(edges.last.sliding$date) ) {
-
+        ## Both, the last sliding network and the last regular network end at the very last edge.
+        ## This is the case because the end of the edges is never cropped (like the beginning is).
+        ## Both split.network.activity.based, and split.network.by.bins, which are invoked to obtain
```
Comment (bockthom) [inline util-split.R:739]: Please use single quotation marks around the functions names here. Please also fix this at the other place from which you have copied this comment :smile:

File: util-split.R [hunk 2]
```diff
@@ -814,6 +814,12 @@ split.network.time.based.by.ranges = function(network, ranges, remove.isolates =
         }
     )
 
+    ## convert ranges to bins
+    bins.starts = sapply(ranges.bounds, function(range) range[1])
+    bins.end = ranges.bounds[[length(ranges.bounds)]][2]
+    bins.date = get.date.from.unix.timestamp(c(bins.starts, bins.end))
```
Comment (bockthom) [inline util-split.R:820]: Shouldn't it be `bins.start` (singular) instead of plural?
Also, I am wondering why this is the only test in which some non-manual bin extraction is performed.
In general, I would be fine with that, but we need to make sure that we do not compute the expected values using the same functions as the ones that are being tested.

File: tests/test-networks-equal-constructions.R [hunk 3]
```diff
@@ -256,7 +256,7 @@ test_that("Compare networks after adding vertex attributes in different order",
 
     ## Order of attributes is now different, while the content is the same.
     ## The resulting networks are therefore not equal.
-    expect_false(compare(networks.both.1, networks.both.2)$equal)
+    expect_false(identical(networks.both.1, networks.both.2))
```
Comment (bockthom) [inline tests/test-networks-equal-constructions.R:259]: Interesting change. I don't know what's going on here. If we keep this change, we definitively need more description in the commit message why this is necessary. However, let's wait for how we solve the igraph-version issue in all the other tests before we deal with this one here...

**Review by bockthom — CHANGES_REQUESTED** [review]
Looks good, thanks for all the fixes and improvements @MaLoefUDS. 
I just spotted a few commenting issues and a two implementation choices where I am not sure if they are correct. Please have a look at my detailed comments.

File: tests/test-networks-equal-constructions.R (hunk 3 above)
Comment (bockthom) [reply tests/test-networks-equal-constructions.R:259]: This change still looks weird to me. Sometimes, the test fails and sometimes not - it is not really reproducible...
So I would like to ignore this one for the moment.

Regarding the other tests that fail because of igraph version 2.0.1.1: The multi-network tests fail because it now is not allowed any more to add an empty set of vertices to a network. I have added a simple check for that, the fix is contained in PR #250. This reduced the number of failing tests to 4 or 5. Four of the remaining tests fail in the motif identification due to a bug in the new igraph version. If have already reported this bug to igraph and they already confirmed it and are on they way to fix it. So, we can ignore those failing motif tests for now until this is fixed in igraph.

What does this mean for our merge queue @hechtlC @MaLoefUDS: I suggest to merge #250 first to get rid of the failing multi-network tests. Afterwards we can merge this PR (as soon as all the issues detected above are fixed), unless of the motif tests are failing or not. Regarding the `compare` vs. `identical` failing test, which only sometimes fails, further investigations are necessary. If your fix really fixes this problem we can keep it, but I am not sure what's going on there yet...

File: util-split.R (hunk 2 above)
Comment (MaLoefUDS) [reply util-split.R:820]: `bins.starts` refers to the start dates of each of the ranges, therefore it is plural. `bins.end` however, only refers to the end date of the very last range, which makes it singular. Also im not sure if I miss something about the tests you mentioned but this change is inside an implementation file, not a test file ^^

File: util-networks.R [hunk 4]
```diff
@@ -486,11 +486,55 @@ NetworkBuilder = R6::R6Class("NetworkBuilder",
...
+                jira.add.links = add.links[add.links$issue.source == "jira", ]
+                matched = c()
+
+                ## iterate over all add_link events from jira
+                for (i in 1:nrow(jira.add.links)) {
+
+                    add.link = jira.add.links[i, ]
+                    if (all(add.link %in% matched)) {
+                        next
+                    }
```
[the remark this answers is not in the document]
Comment (MaLoefUDS) [reply util-networks.R:510]: Point (1) is a good point that I oversaw due to my lack of understanding of R 💀. However, the check cannot simply be removed. The reason it is there is so that not *both* edges are removed in the process. You can observe that not the `add.link` object is added to the `matched` list but its counterpart, `match`. Therefore, I will try to find a better data structure, as suggested in (1)

[the remark this answers is not in the document]
Comment (bockthom) [reply util-networks.R]: Agreed, your new implementation looks good to me.

**Comment by bockthom** [comment]
In general, I am fine with this PR now. However, there is one failing test (but I don't know why the test fails): 

```
Failure ('test-networks-artifact.R:165:5'): Network construction of an issue-based artifact-network directed: FALSE ──
igraph::identical_graphs(network.built, network.expected) is not TRUE

`actual`:   FALSE
`expected`: TRUE 
```

As soon as you manage to fix the above problem, it is time to add the changelog :) For the sliding-window-related bug fixes and for the issue-artifact-network fixes, I suggest to just add the new commits to the already existing items in the changelog - the corresponding fixes you made in this PR are fixes to unreleased changes and, thus, do not belong to the fixes section. (Of course, the additional fixes that do not belong to/follow-up the changes of your previous PR should become individual items in the changelog).

In addition, please don't forget to update copyright headers in the following files:
- test-networks-artifact.R
- test-networks-author.R
- test-networks-equal-constructions.R
- test-read.R
- test-split-data-time-based.R

**Comment by MaLoefUDS** [comment]
The mentioned test fails because of the JIRA issue deduplication. I don't know what happened but it seems like something in R changed. For example, I used `seq_along(jira.add.links)` and previously it used to iterate over rows (i.e., individual events) of the sub-dataframe as expected, but now it iterates over the columns (i.e., `issue.id`, `date` etc.). Similarly I previously used `matched = append(matched, as.list(match))` to add matched events to a list of list, now this does not add every event row as one element in the `matched` list anymore, but adds every individual field (as it was before with `c`, which we discussed in our meeting). Now it only works when I use `list(...)` instead of `as.list(...)`, and I am pretty sure that it was entirely the other way around when I last checked and pushed 💀.

Further, for me I see more failing tests than that one. Following is an error log from my local installation (after fixing the JIRA issue deduplication weirdness):

```R
── Error (test-motifs.R:35:5): Motifs are found in sample network (remove.duplicates = TRUE). ──
Error in `graph.get.subisomorphisms.vf2(target, pattern, ...)`: REAL() can only be applied to a 'numeric', not a 'integer'
Backtrace:
    ▆
 1. └─global motifs.search.in.network(network, MOTIFS.LINE, remove.duplicates = TRUE) at test-motifs.R:35:5
 2.   └─igraph::subgraph_isomorphisms(...) at util-motifs.R:136:5
 3.     └─igraph::graph.get.subisomorphisms.vf2(target, pattern, ...)
── Error (test-motifs.R:89:5): Motifs are found in sample network (remove.duplicates = FALSE). ──
Error in `graph.get.subisomorphisms.vf2(target, pattern, ...)`: REAL() can only be applied to a 'numeric', not a 'integer'
Backtrace:
    ▆
 1. └─global motifs.search.in.network(network, MOTIFS.LINE, remove.duplicates = FALSE) at test-motifs.R:89:5
 2.   └─igraph::subgraph_isomorphisms(...) at util-motifs.R:136:5
 3.     └─igraph::graph.get.subisomorphisms.vf2(target, pattern, ...)
── Error (test-motifs.R:165:5): Motif counts for sample network (remove.duplicates = TRUE). �─
�Error in `graph.get.subisomorphisms.vf2(target, pattern, ...)`: REAL() can only be applied to a 'numeric', not a 'integer'
Backtrace:
    ▆
 1. └─global motifs.count(...) at test-motifs.R:165:5
 2.   └─base::lapply(...) at util-motifs.R:202:5
 3.     └─FUN(X[[i]], ...)
 4.       └─global motifs.search.in.network(network = network, motif = motif, remove.duplicates = remove.duplicates) at util-motifs.R:203:9
 5.         └─igraph::subgraph_isomorphisms(...) at util-motifs.R:136:5
 6.           └─igraph::graph.get.subisomorphisms.vf2(target, pattern, ...)
── Error (test-networks.R:49:5): Simplify network with more than one relation ──
Error in `add_vertices(e1, la, attr = e2)`: At rinterface_extra.c:78 : Expecting a scalar integer but received a vector of length 2. Invalid value
Backtrace:
    ▆
 1. └─igraph:::`+.igraph`(...) at test-networks.R:49:5
 2.   └─igraph::add_vertices(e1, la, attr = e2)

[ FAIL 4 | WARN 71 | SKIP 0 | PASS 2504 ]
```

**Comment by bockthom** [comment]
> The mentioned test fails because of the JIRA issue deduplication. I don't know what happened but it seems like something in R changed. 

I don't think so. There was no R update in the last months (to the best of my knowledge), and also older R versions are affected. Maybe something in your local installation was broken beforehand. Anyway, if you already have identified why the test has failed, then it should be easily fixable, I guess.

> Further, for me I see more failing tests than that one. 

Which igraph version do you use locally? All the four failing tests have been failing in igraph 2.0.1.1 due to a newly introduced bug in igraph. However, this bug has been fixed in 2.0.2. So, when you use igraph 2.0.2, these four tests should not fail any more (as they also didn't fail in the last CI run here).

**Comment by MaLoefUDS** [comment]
Hmm still weird, I changed nothing in my configuration so I have no idea where this came from.

I was on `igraph` 2.0.1.1, bumping to 2.0.2 fixed the remaining bugs 👍.

**Comment by MaLoefUDS** [comment]
If you were wondering about all the force pushes right now, I removed the one commit which only changed a single test for igraph version 2.0.1.1. I found that the change is not necessary anymore.

File: NEWS.md [hunk 5]
```diff
@@ -11,23 +11,27 @@
...
 - Enhance `get.author.names.from.network` and `get.author.names.from.data` to always have the same output format. Now it doesn't depend on the `global` flag anymore (PR #248, d87d32564156f13c83ebe3361c2b68e5d0ac16ac, ddbfe68d3e628e82f34e09b36fffe886646986c5)
 - Change `util-tensor.R` to correctly use the new output format of `get.author.names.from.network` (PR #248, 72b663ebf7169c0da5c687fe215529f3be0c08c5)
 - Throw an error in `convert.adjacency.matrix.list.to.array` if the function is called with wrong parameters (PR #248, ece2d38b4972745af3a83e06f32317a06465a345, 1a3e510df15f5fa4e920e9fce3e0e162c27cd6d1)
 - Rename `compare.networks` to `assert.networks.equal` to better match the purpose of the function (PR #248, d9f1a8d90e00a634d7caeb5e7f8f262776496838)
 - Explicitly add R version 4.3 to the CI test pipeline (9f346d5bc3cfc553f01e5e80f0bbe51e1dc2b53e)
+- If possible, disambiguate duplicate JIRA issue-edges resulting from the codeface extraction (PR #249, fa3167c289c9785f3a5db03d9724848f1441a63d, 4646d581d5e1f63260692b396a8bd8f51b0da48fda, ed77bd726bf92e06c2fc9145a5847787a8d0588b)
```
Comment (bockthom) [inline NEWS.md:27]: There are no issue artifact networks yet, so this is nothing which can be improved or changed. As this still belongs to the initial implementation of issue artifact networks, dedublication of JIRA issue-edges needs not to be mentioned separately. Please move this item to the very first "Added" item and merge the two items into one item.

File: NEWS.md [hunk 6]
```diff
@@ -11,23 +11,27 @@
 - Add new `assert.sparse.matrices.equal` function to compare two sparse matrices for equality for testing purposes (PR #248, 9784cdf12d1497ee122e2ae73b768b8c334210d4, d9f1a8d90e00a634d7caeb5e7f8f262776496838)
 - Add tests for file `util-networks.misc.R` for issue #242 (PR #248, f3202a6f96723d11c170346556d036cf087521c8, 030574b9d0f3435db4032d0e195a3d407fb7244b, 380b02234275127297fcd508772c69db21c216de, 8b803c50d60fc593e4e527a08fd4c2068d801a48, 7335c3dd4d0302b024a66d18701d9800ed3fe806, 6b600df04bec1fe70c272604f274ec5309840e65)
 - Add the possibility to simplify edges of multiple-relation networks into a single edge at all instead of a single edge per relation (PR #250, 2105ea89b5227e7c9fa78fea9de1977f2d9e8faa)
+- Add `get.bin.dates.from.ranges` function to convert date ranges into bins format (PR #249, a1842e9be46596321ee86860fd87d17a3c88f50f)
+- Add tests for the presence and validity of the `bins` attribute on network-, and data-splits (PR #249, c064affcfff2eb170d8bdcb39d837a7ff62b2cbd, 93051ab848ec94de138b0513dac22f6da0d20885)
```
Comment (bockthom) [inline NEWS.md:15]: This is nothing that has been newly added. Instead, you improved already existing tests, so I would move this item to the "Changed/Improved" section. I would suggest to also start with "Test for the presence..." (or something similar) instead of "Add tests" as you didn't add any test but just added additional checks to the existing tests.

File: NEWS.md [hunk 7]
```diff
@@ -11,23 +11,27 @@
...
 - Change `util-tensor.R` to correctly use the new output format of `get.author.names.from.network` (PR #248, 72b663ebf7169c0da5c687fe215529f3be0c08c5)
 - Throw an error in `convert.adjacency.matrix.list.to.array` if the function is called with wrong parameters (PR #248, ece2d38b4972745af3a83e06f32317a06465a345, 1a3e510df15f5fa4e920e9fce3e0e162c27cd6d1)
 - Rename `compare.networks` to `assert.networks.equal` to better match the purpose of the function (PR #248, d9f1a8d90e00a634d7caeb5e7f8f262776496838)
 - Explicitly add R version 4.3 to the CI test pipeline (9f346d5bc3cfc553f01e5e80f0bbe51e1dc2b53e)
+- If possible, disambiguate duplicate JIRA issue-edges resulting from the codeface extraction (PR #249, fa3167c289c9785f3a5db03d9724848f1441a63d, 4646d581d5e1f63260692b396a8bd8f51b0da48fda, ed77bd726bf92e06c2fc9145a5847787a8d0588b)
+- Improve overall consistency in network-splitting code (PR #249, a1842e9be46596321ee86860fd87d17a3c88f50f)
```
Comment (bockthom) [inline NEWS.md:28]: Could you please be a little bit more precise what "improve overall consistency" means? Keep in mind that a user reads this and needs to be able to judge whether this change affects the outcome of the application or usage of network splitting...

File: NEWS.md [hunk 8]
```diff
@@ -11,23 +11,27 @@
 - Add new `assert.sparse.matrices.equal` function to compare two sparse matrices for equality for testing purposes (PR #248, 9784cdf12d1497ee122e2ae73b768b8c334210d4, d9f1a8d90e00a634d7caeb5e7f8f262776496838)
 - Add tests for file `util-networks.misc.R` for issue #242 (PR #248, f3202a6f96723d11c170346556d036cf087521c8, 030574b9d0f3435db4032d0e195a3d407fb7244b, 380b02234275127297fcd508772c69db21c216de, 8b803c50d60fc593e4e527a08fd4c2068d801a48, 7335c3dd4d0302b024a66d18701d9800ed3fe806, 6b600df04bec1fe70c272604f274ec5309840e65)
 - Add the possibility to simplify edges of multiple-relation networks into a single edge at all instead of a single edge per relation (PR #250, 2105ea89b5227e7c9fa78fea9de1977f2d9e8faa)
+- Add `get.bin.dates.from.ranges` function to convert date ranges into bins format (PR #249, a1842e9be46596321ee86860fd87d17a3c88f50f)
```
Comment (bockthom) [inline NEWS.md:14]: Please also add `858b1812ebfc3194cc6a03c99f3ee7d161d1ca15` here as this commits moves the function to another file. Both commits together account for adding the file at the location at which it is placed at the end of the PR.

**Review by bockthom — CHANGES_REQUESTED** [review]
I have added a few comments regarding the changes to the changelog.

**Review by bockthom — APPROVED** [review]

**Comment by bockthom** [comment]
I will merge right away, although the CI pipeline currently fails in the showcase file for R versions 3.6 and 4.0. This is caused by a build failure in a new release of package `ggraph` that was released yesterday. I contacted the maintainer of `ggraph` regarding that and he is currently about to find a solution for this problem, so this is nothing we should care about (for now).
